File: examples/core/read_sensor.py
#!/usr/bin/env python3
"""
Read AHT21 sensor data from Johnny552 via HID INPUT reports.
"""

import hid
import struct
import time
import sys

# USB VID:PID for Johnny552
VID = 0x16C0
PID = 0x5552

def parse_sensor_data(data):
    """Parse the 8-byte HID INPUT report containing sensor data."""
    if len(data) < 8:
        return None, None
    
    # Report format: [ReportID, Temp_H, Temp_L, Hum_H, Hum_L, 0, 0, 0]
    report_id = data[0]
    
    if report_id != 0x01:
        return None, None
    
    # Extract temperature (signed 16-bit, °C × 10)
    temp_raw = (data[1] << 8) | data[2]
    # Convert to signed
    if temp_raw & 0x8000:
        temp_raw -= 0x10000
    temperature = temp_raw / 10.0
    
    # Extract humidity (unsigned 16-bit, % × 10)
    hum_raw = (data[3] << 8) | data[4]
    humidity = hum_raw / 10.0
    
    return temperature, humidity

def main():
    """Main function to read sensor data continuously."""
    print(f"Looking for Johnny552 device (VID: {VID:04X}, PID: {PID:04X})...")
    
    try:
        # Open the HID device
        device = hid.device()
        device.open(VID, PID)
        
        print(f"Connected to: {device.get_manufacturer_string()} - {device.get_product_string()}")
        print("Reading sensor data... (Press Ctrl+C to exit)\n")
        
        # Set non-blocking mode with timeout
        device.set_nonblocking(False)
        
        while True:
            # Read with timeout (1000ms)
            data = device.read(8, timeout_ms=1000)
            
            if data:
                # Parse the sensor data
                temperature, humidity = parse_sensor_data(data)
                
                if temperature is not None and humidity is not None:
                    print(f"Temperature: {temperature:5.1f}°C | Humidity: {humidity:4.1f}%")
                    sys.stdout.flush()
            
            # Small delay
            time.sleep(0.1)
    
    except IOError as e:
        print(f"Error: Could not open device. {e}")
        print("Make sure the device is connected and you have permissions.")
        print("You may need to run: sudo python3 read_sensor.py")
        return 1
    except KeyboardInterrupt:
        print("\nExiting...")
        return 0
    finally:
        try:
            device.close()
        except:
            pass
    
    return 0

if __name__ == "__main__":
    sys.exit(main())

File: examples/core/read_sensor.c
/*
 * Read AHT21 sensor data from Johnny552 via HID
 * Compile: gcc -o read_sensor read_sensor.c -lhidapi-libusb
 * Run: sudo ./read_sensor
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <hidapi/hidapi.h>

#define VID 0x16C0
#define PID 0x5552

static int running = 1;

void signal_handler(int sig) {
    running = 0;
}

int main() {
    hid_device *handle;
    unsigned char buf[8];
    int res;
    
    // Initialize signal handler
    signal(SIGINT, signal_handler);
    
    // Initialize HIDAPI
    if (hid_init() < 0) {
        fprintf(stderr, "Error: Failed to initialize HIDAPI\n");
        return 1;
    }
    
    // Open the device
    handle = hid_open(VID, PID, NULL);
    if (!handle) {
        fprintf(stderr, "Error: Unable to open device %04X:%04X\n", VID, PID);
        fprintf(stderr, "Try running with sudo\n");
        hid_exit();
        return 1;
    }
    
    printf("Connected to Johnny552\n");
    printf("Reading sensor data... (Press Ctrl+C to exit)\n\n");
    
    // Set non-blocking mode
    hid_set_nonblocking(handle, 0);
    
    while (running) {
        // Read with 1 second timeout
        res = hid_read_timeout(handle, buf, sizeof(buf), 1000);
        
        if (res < 0) {
            fprintf(stderr, "Error reading: %ls\n", hid_error(handle));
            break;
        }
        
        if (res > 0) {
            // Parse sensor data
            if (buf[0] == 0x01 && res >= 5) {
                // Extract temperature (signed 16-bit)
                int16_t temp_raw = (buf[1] << 8) | buf[2];
                float temperature = temp_raw / 10.0f;
                
                // Extract humidity (unsigned 16-bit)
                uint16_t hum_raw = (buf[3] << 8) | buf[4];
                float humidity = hum_raw / 10.0f;
                
                printf("Temperature: %5.1f°C | Humidity: %4.1f%%\n", temperature, humidity);
                fflush(stdout);
            }
        }
        
        usleep(100000);  // 100ms delay
    }
    
    printf("\nExiting...\n");
    hid_close(handle);
    hid_exit();
    
    return 0;
}
